discovery-provider/src/tasks/index_rewards_manager.py

```diff
@@ -0,0 +1,369 @@
+                # Current batch of transactions
+                transaction_signature_batch = []
+                for tx_info in transactions_array:
+                    tx_sig = tx_info["signature"]
+                    tx_slot = tx_info["slot"]
+                    logger.info(
+                        f"index_rewards_manager.py | Processing tx={tx_sig} | slot={tx_slot}"
+                    )
+                    if tx_info["slot"] > latest_processed_slot:
+                        transaction_signature_batch.append(tx_sig)
+                    elif tx_info["slot"] <= latest_processed_slot and (
+                        min_slot is None or tx_info["slot"] > min_slot
+                    ):
+                        # Check the tx signature for any txs in the latest batch,
+                        # and if not present in DB, add to processing
+                        logger.info(
+                            f"index_rewards_manager.py | Latest slot re-traversal\
+                            slot={tx_slot}, sig={tx_sig},\
+                            latest_processed_slot(db)={latest_processed_slot}"
+                        )
+                        exists = check_tx_exists(session, tx_sig)
+                        if exists:
+                            intersection_found = True
+                            break
+                        # Ensure this transaction is still processed
+                        transaction_signature_batch.append(tx_sig)
+
+                # Restart processing at the end of this transaction signature batch
+                last_tx = transactions_array[-1]
+                last_tx_signature = last_tx["signature"]
+
+                # Append batch of processed signatures
+                if transaction_signature_batch:
+                    transaction_signatures.append(transaction_signature_batch)
+
+                # Ensure processing does not grow unbounded
+                if len(transaction_signatures) > TX_SIGNATURES_MAX_BATCHES:
+                    # Only take the oldest transaction from the transaction_signatures array
+                    # transaction_signatures is sorted from newest to oldest
+                    transaction_signatures = transaction_signatures[
+                        -TX_SIGNATURES_RESIZE_LENGTH:
+                    ]
+
+    # Reverse batches aggregated so oldest transactions are processed first
+    transaction_signatures.reverse()
+    return transaction_signatures
+
+
+def process_transaction_signature(
+    solana_client: Client, db: SessionManager, transaction_signatures: List[str]
+):
+    """Concurrently processes the transactions to update the DB state for reward transfer instructions"""
+    for tx_sig_batch in transaction_signatures:
+        logger.info(f"index_rewards_manager.py | processing {tx_sig_batch}")
+        batch_start_time = time.time()
+        # Process each batch in parallel
+        with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
```

Comment:
Let's watch the perf of this really carefully. I'm not in love with each executor making multiple RTs to the DB for each disbursement. This could be rewritten to collect the txn info from solana first, and then do a single DB roundtrip to get all of those users referenced in those disbursements, rather than one RT per disbursement

Reply:
Interesting call out. I know we do this in indexing solana plays which I believe has much more volume - any thoughts on perf we've seen from that @hareeshnagaraj

Reply:
It's a good call out - the indexing sol plays test writes quite a bit of volume but it is very isolated compared to a prod environment. let's chat offline about this w/@piazzatron  might be worth optimizing in this sprint